### PR TITLE
 Autocomplete Options in flowR repl

### PR DESCRIPTION
**Pull request by Ellpeck**

### CODE REVIEW

File: src/cli/common/scripts-info.ts [hunk 1]
```diff
@@ -99,3 +100,14 @@ const _scripts = {
 
 export const scripts = _scripts as Record<keyof typeof _scripts, ScriptInformation>
 
+export function getValidOptionsForCompletion(options: OptionDefinition[], prevArgs: string[]): string[] {
+	return options
+		.filter(o => o.multiple || !prevArgs.includes(asOptionName(o.name)) && (!o.alias || !prevArgs.includes(asOptionName(o.alias))))
```
Comment (EagleoutIce) [inline src/cli/common/scripts-info.ts:105]: i think we can make this more readable by outsourcing the filter to a function with a nice name!

File: src/cli/repl/core.ts [hunk 2]
```diff
@@ -8,12 +8,14 @@ import { bold } from '../../statistics'
 import { prompt } from './prompt'
 import type { ReplOutput } from './commands'
 import { commandNames, getCommand, standardReplOutput } from './commands'
+
```
Comment (EagleoutIce) [inline src/cli/repl/core.ts:11]: space me :sparkles:

Comment (Ellpeck) [reply src/cli/repl/core.ts:11]: i actually think that one weird eslint plugin rule thing we added a while back to combine imports occasionally puts empty lines for seemingly no reason ✨ truly a technological wonder

File: src/cli/common/scripts-info.ts [hunk 3]
```diff
@@ -99,3 +100,16 @@ const _scripts = {
 
 export const scripts = _scripts as Record<keyof typeof _scripts, ScriptInformation>
 
+export function getValidOptionsForCompletion(options: OptionDefinition[], prevArgs: string[]): string[] {
```
Comment (EagleoutIce) [inline src/cli/common/scripts-info.ts:103]:
```suggestion
export function getValidOptionsForCompletion(options: readonly OptionDefinition[], prevArgs: readonly string[]): string[] {
```

File: src/cli/common/scripts-info.ts [hunk 4]
```diff
@@ -99,3 +100,16 @@ const _scripts = {
...
+	return options.filter(o => canAddOption(o, prevArgs)).flatMap(o => {
+		const args = [asOptionName(o.name)]
+		if(o.alias) {
+			args.push(asOptionName(o.alias))
+		}
+		return args
+	})
+}
+
+function canAddOption(option: OptionDefinition, prevArgs: string[]): boolean {
```
Comment (EagleoutIce) [inline src/cli/common/scripts-info.ts:113]:
```suggestion
function canAddOption(option: OptionDefinition, prevArgs: readonly string[]): boolean {
```

File: src/cli/repl/core.ts [hunk 5]
```diff
@@ -22,7 +24,36 @@ const defaultHistoryFile = path.join(os.tmpdir(), '.flowrhistory')
...
+			const completions: string[] = []
+
+			const commandName = commandNameColon.slice(1)
+			if(getCommand(commandName)?.script === true){
+				// autocomplete script arguments
+				const options = scripts[commandName as keyof typeof scripts].options
+				completions.push(...getValidOptionsForCompletion(options, splitLine).map(o => `${o} `))
+			} else {
+				// autocomplete command arguments (specifically, autocomplete the file:// protocol)
+				completions.push('file://')
```
Comment (EagleoutIce) [inline src/cli/repl/core.ts:44]: maybe we can define this magic prefix as a global constant just so it is not repeated too often? :pretty-please:

**Comment by Ellpeck** [comment]
i dare you to reject again

**Review by EagleoutIce — APPROVED** [review]

**Comment by EagleoutIce** [comment]
_This pull request is included in v1.4.2 (see [Release v1.4.2 (Dropping xmlparsedata, Benchmark Re-Runs, and Repl Fixes)](https://github.com/Code-Inspect/flowr/releases/tag/v1.4.2))._

**Pull request merged**